### ISSUE TITLE
Ensure Serializable Security Components declare serialVersionUID

### ISSUE THREAD

**Issue by jzheaux** [spring-projects/spring-security#16276]
To ensure backward compatibility, Security components that implement `Serializable` should have a `serialVersionUID`.

Based on internal testing across a few dozen JVMs, it appears that the `serialVersionUID` is consistent for Security's components. As such, we can safely add the calculated `serialVersionUID` value to each class that is missing it during the 6.4.x maintenance cycle.

- [ ] #16274 
- [x] Produce list of `Serializable` classes missing `serialVersionUID`
- [ ] Fail build when a `Serializable` class is missing a `serialVersionUID`
- [ ] Consider adding an exclusion list for classes we don't support being serialized

When addressing a class that is missing its `serialVersionUID`, please do the following:

1. Add the calculated `serialVersionUID` (IDEs can usually do this for you, or you can use `serialver` which ships with the JVM)
2. In `SpringSecurityCoreVersionSerializableTests`, add the class and an example construction to the `generatorByClassName` map
3. Run `SpringSecurityCoreVersionSerializableTests#serializeCurrentVersionClasses`. 
4. If successful, it will create a `{className}.serialized` file in `config/src/main/resources/serialized`:

    Run the other tests in `SpringSecurityCoreVersionSerializableTests`; the class should be added to the list in `shouldBeAbleToDeserializeClassFromPreviousVersion`, the class should no longer be in the output for `listClassesMissingSerialVersion`

    Commit the `Serialiizable` class(es) and `SpringSecurityCoreVersionSerializableTests`
5. If unsuccessful, it is usually because one of its members is not serializable. Find the unserializable member; file a ticket to ensure that it is made `Serializable`

You can also see the list of `Serializable` files by running:
```bash
./gradlew :spring-security-config:test --tests "*MissingSerialVersion*" -Pserialization
```

**Comment by jzheaux** [comment]
Here is an example: https://github.com/spring-projects/spring-security/commit/e3cd4339b2c5cb43f5cde41b7ecbd659a1d8bf59

**Comment by jzheaux** [comment]
I'll take care of the following classes:

```
ldap/src/main/java/org/springframework/security/ldap/userdetails/LdapUserDetailsManager.java
oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/introspection/SpringOpaqueTokenIntrospector.java
oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/introspection/SpringReactiveOpaqueTokenIntrospector.java
config/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/client/OidcBackChannelLogoutAuthentication.java
config/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/client/OidcLogoutAuthenticationToken.java
config/src/main/java/org/springframework/security/config/web/server/OidcBackChannelLogoutAuthentication.java
config/src/main/java/org/springframework/security/config/web/server/OidcLogoutAuthenticationToken.java
config/src/test/java/org/springframework/security/SpringSecurityCoreVersionSerializableTests.java
core/src/main/java/org/springframework/security/authentication/ott/OneTimeTokenAuthenticationToken.java
ldap/src/main/java/org/springframework/security/ldap/userdetails/LdapAuthority.java
oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/logout/OidcLogoutToken.java
oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/session/OidcSessionInformation.java
oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/DefaultOAuth2AuthenticatedPrincipal.java
oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/OAuth2AccessToken.java
oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/OAuth2DeviceCode.java
oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/OAuth2RefreshToken.java
oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/OAuth2UserCode.java
oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/oidc/OidcIdToken.java
oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/oidc/user/DefaultOidcUser.java
oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/oidc/user/OidcUserAuthority.java
oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/Jwt.java
oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/BearerTokenError.java
oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/BearerTokenAuthentication.java
oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/introspection/OAuth2IntrospectionAuthenticatedPrincipal.java
saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/DefaultSaml2AuthenticatedPrincipal.java
saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/Saml2Authentication.java
saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/Saml2AuthenticationToken.java
saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/Saml2PostAuthenticationRequest.java
saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/Saml2RedirectAuthenticationRequest.java
saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/logout/Saml2LogoutRequest.java
```

My intention is to target authentication tokens, credentials, and authorization objects.

**Comment by dalbani** [comment]
> Based on internal testing across a few dozen JVMs, it appears that the `serialVersionUID` is consistent for Security's components. As such, we can safely add the calculated `serialVersionUID` value to each class that is missing it during the 6.4.x maintenance cycle.

Upgrading my application storing sessions in database from Spring Boot 3.3.6 to Spring Boot 3.4.1 (i.e. Spring Security 6.4.2), I ran into this issue:
```
java.io.InvalidClassException: org.springframework.security.oauth2.core.oidc.user.OidcUserAuthority; local class incompatible: stream classdesc serialVersionUID = -4675866280835753141, local class serialVersionUID = -196018737016047617
```

So it looks like that the `serialVersionUID` that my JVM (RedHat's OpenJDK 21.0.5+10-LTS) calculated is different that yours, right?

I'm surprised that no one has been hit by the same issue 🤔

**Comment by jzheaux** [comment]
Hi, @dalbani, thanks for the report. I'm a little surprised to see a difference as we internally tested several dozen JVMs to check the calculated value.

That may be why no one else has reported just yet. 

We need to set a value at some point; otherwise folks will break every time one of these classes is edited.  I'll take a look at the JVM you listed and get back to you.

**Comment by jzheaux** [comment]
Here is my output from [the latest from RedHat](https://developers.redhat.com/content-gateway/file/openjdk/21.0.5.0.11/java-21-openjdk-21.0.5.0.11-1.portable.jdk.x86_64.tar.xz):

```
~/Downloads/java-21-openjdk-21.0.5.0.11-1.portable.jdk.x86_64/bin/serialver \
-classpath oauth2/oauth2-core/build/libs/spring-security-oauth2-core-6.4.2-SNAPSHOT.jar:core/build/libs/spring-security-core-6.4.2-SNAPSHOT.jar \
org.springframework.security.oauth2.core.oidc.user.OidcUserAuthority
org.springframework.security.oauth2.core.oidc.user.OidcUserAuthority:    private static final long serialVersionUID = -4675866280835753141L;
```

This appears to agree with the value in `OidcUserAuthority` currently. 

However, if I look at older versions:

#### Spring Security 6.2.7

```bash
~/Downloads/java-21-openjdk-21.0.5.0.11-1.portable.jdk.x86_64/bin/serialver \
-classpath oauth2/oauth2-core/build/libs/spring-security-oauth2-core-6.2.7-SNAPSHOT.jar:core/build/libs/spring-security-core-6.2.7-SNAPSHOT.jar \
org.springframework.security.oauth2.core.oidc.user.OidcUserAuthority
org.springframework.security.oauth2.core.oidc.user.OidcUserAuthority:    private static final long serialVersionUID = -196018737016047617L;

```

#### Spring Security 6.3.0

```bash
~/Downloads/java-21-openjdk-21.0.5.0.11-1.portable.jdk.x86_64/bin/serialver \
-classpath oauth2/oauth2-core/build/libs/spring-security-oauth2-core-6.3.0-SNAPSHOT.jar:core/build/libs/spring-security-core-6.3.0-SNAPSHOT.jar \
org.springframework.security.oauth2.core.oidc.user.OidcUserAuthority
org.springframework.security.oauth2.core.oidc.user.OidcUserAuthority:    private static final long serialVersionUID = -196018737016047617L;

```

Then I think it's a little clearer what's going on.

Because `OidcUserAuthority` [changed its public API in 6.4](https://github.com/spring-projects/spring-security/commit/99aee99b3457a1ecebc05d543fb417a0b0008d11), this resulted in a different calculated `serialVersionUID`. The local id you are seeing is likely from a serialized version of `OidcUserAuthority` from a previous minor version of Spring Security.

In other words, this is something you would have experienced when upgrading, independent of the commit that added `serialVersionUID` to this class. Once this ticket is complete, then deserializations across future minor versions of Spring Security will be much more stable.

As a whole, this is something that we are addressing in order [to eliminate deserialization issues across minor versions in Spring Security](https://docs.spring.io/spring-security/reference/6.3/whats-new.html#_passive_jdk_serialization_support). You can read https://github.com/spring-projects/spring-security/issues/16163 to learn about the issue that triggered the need for this specific ticket.

**Comment by dalbani** [comment]
Thanks @jzheaux for your feedback.
The issue comes indeed from a different `serialVersionUID` value, when session objects originally serialized with Spring Security pre-6.4 are then read with Spring Security 6.4.

I was actually aware of the effort mentioned in https://docs.spring.io/spring-security/reference/6.3/whats-new.html#_passive_jdk_serialization_support  — and I very much appreciate the improvement.

Though I'm not sure this "concern [has become] a thing of the past" as mentioned on https://spring.io/blog/2024/01/19/spring-security-6-3-adds-passive-jdk-serialization-deserialization-for.
I mean, are all Java classes now covered? Won't there be new changes in say Spring Security 6.5?

I suppose I need to ask the people in charge of Spring Session, but what do think of introducing a behaviour in Spring Session that failing to deserialize a session automatically invalidates it?
So that manually messing with sessions stored in database is not necessary anymore when doing "breaking" upgrades of Spring Security.

**Comment by jzheaux** [comment]
You are correct that @dalbani was perhaps too confldent in hindsight. At the time, there was not an analysis performed to see which classes were missing an id.

I think this point in the description:

```
Fail build when a Serializable class is missing a serialVersionUID
```

Will go a long way to catching these before future releases.

**Comment by sn3jkeee** [comment]
Experienced same issues as @dalbani mentioned, has to clear out Redis as it used by Spring Session. Quite unfortunate all users forced to logged out
